polars/polars-io/src/ipc/ipc_stream.rs

```diff
@@ -0,0 +1,486 @@
+//!
+//! // Create an in memory file handler.
+//! // Vec<u8>: Read + Write
+//! // Cursor<T>: Seek
+//!
+//! let mut buf: Cursor<Vec<u8>> = Cursor::new(Vec::new());
+//!
+//! // write to the in memory buffer
+//! IpcStreamWriter::new(&mut buf).finish(&mut df).expect("ipc writer");
+//!
+//! // reset the buffers index after writing to the beginning of the buffer
+//! buf.set_position(0);
+//!
+//! // read the buffer into a DataFrame
+//! let df_read = IpcStreamReader::new(buf).finish().unwrap();
+//! assert!(df.frame_equal(&df_read));
+//! ```
+use crate::predicates::PhysicalIoExpr;
+use crate::{finish_reader, ArrowReader, ArrowResult};
+use crate::{prelude::*, WriterFactory};
+use arrow::io::ipc::write::WriteOptions;
+use arrow::io::ipc::{read, write};
+use polars_core::prelude::*;
+
+use std::io::{Read, Seek, Write};
+
+use std::path::PathBuf;
+use std::sync::Arc;
+
+/// Read Arrows Stream IPC format into a DataFrame
+///
+/// # Example
+/// ```
+/// use polars_core::prelude::*;
+/// use std::fs::File;
+/// use polars_io::ipc::IpcStreamReader;
+/// use polars_io::SerReader;
+///
+/// fn example() -> Result<DataFrame> {
+///     let file = File::open("file.ipc").expect("file not found");
+///
+///     IpcStreamReader::new(file)
+///         .finish()
+/// }
+/// ```
+#[must_use]
+pub struct IpcStreamReader<R> {
+    /// File or Stream object
+    reader: R,
+    /// Aggregates chunks afterwards to a single chunk.
+    rechunk: bool,
+    n_rows: Option<usize>,
+    projection: Option<Vec<usize>>,
+    columns: Option<Vec<String>>,
+    row_count: Option<RowCount>,
+}
+
+impl<R: Read + Seek> IpcStreamReader<R> {
+    /// Get schema of the Ipc Stream File
+    pub fn schema(&mut self) -> Result<Schema> {
+        let metadata = read::read_stream_metadata(&mut self.reader)?;
+        Ok((&metadata.schema.fields).into())
+    }
+
+    /// Get arrow schema of the Ipc Stream File, this is faster than creating a polars schema.
+    pub fn arrow_schema(&mut self) -> Result<ArrowSchema> {
+        let metadata = read::read_stream_metadata(&mut self.reader)?;
+        Ok(metadata.schema)
+    }
+    /// Stop reading when `n` rows are read.
+    pub fn with_n_rows(mut self, num_rows: Option<usize>) -> Self {
+        self.n_rows = num_rows;
+        self
+    }
+
+    /// Columns to select/ project
+    pub fn with_columns(mut self, columns: Option<Vec<String>>) -> Self {
+        self.columns = columns;
+        self
+    }
+
+    /// Add a `row_count` column.
+    pub fn with_row_count(mut self, row_count: Option<RowCount>) -> Self {
+        self.row_count = row_count;
+        self
+    }
+
+    /// Set the reader's column projection. This counts from 0, meaning that
+    /// `vec![0, 4]` would select the 1st and 5th column.
+    pub fn with_projection(mut self, projection: Option<Vec<usize>>) -> Self {
+        self.projection = projection;
+        self
+    }
+
+    // todo! hoist to lazy crate
```

Comment:
This function can be removed as we don't support IPC streams in lazy and we should not do that at this point yet.